existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,303 @@
+* The same variable can't be quantified more than once.
+* You can see that constraints for `b` are added in the middle of the declaration of the function type (because `b` was introduced in the middle of the function).
+* You can choose type for `b` only inside `func` definition, and you can't do it when you call `func` (we will discuss it further below).
+
+Here's a real-life example of using existential types in functions.
+Imagine a function that prints some logs while calculating the result.
+You may want to write logs into a console or into some file.
+In order to do it, you pass a logging function as an argument.
+
+```haskell
+func :: _ -> IO ()
+func log = do
+  username <- getUsername
+  log username
+
+  userCount <- getUserCount
+  log userCount
+
+getUsername :: IO String
+getUsername = undefined
+
+getUserCount :: IO Int
+getUserCount = undefined
+```
+
+What type should the log function have?
+Since it somehow logs both `String` and `Int`:
+
+```haskell
+log :: Show a => a -> IO ()
+```
+However, if we add this signature to `func` and try to call it with different arguments:
+
+```haskell
+func :: Show a => (a -> IO ()) -> IO ()
+func log = ...
+
+main :: IO ()
+main = do
+  -- log to the console
+  func print
+
+  -- log to a logfile
+  func (appendFile "logfile.log" . show)
+```
+We will see some nasty compiler errors:
+```
+Couldn't match expected type `a` with actual type `String`...
+```
+It is happening, because inside of `func`, `a` is not known yet. I will be known, only when you call `func` somewhere.
+The only solution here is to add constraints and `forall` quantifier to the log function signature:
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
+After this change, we can specify.
+
+With this change we can instantiate `a` when we call log.
+
+```haskell
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+func log = do
+  username <- getUsername
+  log @String username
+
+  userCount <- getUserCount
+  log @Int userCount
+```
+
+Note, that this type applications is unnecessary.
+Compiler is smart enough to choose appropriate types.
+
+#### Hiding type variables
+
+Let's turn to datatypes now.
+You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of equality sign.
+
+```haskell
+data MyData a = forall b. MkMyData b a
+-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
+```
+
+The first thing you might have noticed is that we don't have a type variable for `b` on the left side of the expression.
+However, we still can use it on the right side.
+This is because `b` is existentially quantified.
+The compiler will instantiate the type of the variable itself, and you can't specify it.
```

Comment:
Hmm this doesn't sound right... You can absolutely instantiate `b` when you're constructing a `MyData`:

```
md :: MyData Int
md = MkMyData @Int @String "" 1
```

Reply:
Removed MyData part